src/network/mining_pool.rs

```diff
@@ -0,0 +1,324 @@
+                    }
+
+                    // Update miner info for this miner.
+                    let mut miner_info = *info.get(&miner_address).expect("miner should have existing info");
+                    miner_info.0 = chrono::Utc::now().timestamp();
+                    miner_info.2 += 1;
+                    info.insert(miner_address, miner_info);
+
+                    // Since a worker will swap out the difficulty target for their share target,
+                    // let's put it back to the original value before checking the POSW for true
+                    // validity.
+                    let difficulty_target = current_template.difficulty_target;
+                    block.set_difficulty_target(difficulty_target);
+
+                    // If the block is valid, broadcast it.
+                    if block.is_valid() {
+                        debug!("Mining pool has found unconfirmed block {} ({})", block.height(), block.hash());
+
+                        // Store coinbase record(s)
+                        records.iter().for_each(|r| {
+                            if let Err(error) = self.state.add_coinbase_record(block.height(), r.clone()) {
+                                warn!("Could not store coinbase record {}", error);
+                            }
+                        });
+
+                        // Broadcast the next block.
+                        let request = LedgerRequest::UnconfirmedBlock(self.local_ip, block, self.prover_router.clone());
+                        if let Err(error) = self.ledger_router.send(request).await {
+                            warn!("Failed to broadcast mined block - {}", error);
+                        }
+                    }
+                } else {
+                    warn!("[ProposedBlock] No current template exists");
+                }
+            }
+            MiningPoolRequest::BlockHeightClear(block_height) => {
+                // Remove the shares for the given block height.
+                if let Err(error) = self.state.remove_shares(block_height) {
+                    warn!("[BlockHeightClear] {}", error);
+                }
+            }
+            MiningPoolRequest::GetCurrentBlockTemplate(peer_ip, address) => {
+                if let Some(current_template) = &*self.current_template.read().await {
+                    // Ensure this miner exists in the info list first, so we can get their share
+                    // difficulty.
+                    if let None = self.miner_info.read().await.get(&address) {
+                        // If not, let's add them.
+                        let share_difficulty = current_template.difficulty_target.saturating_mul(50);
+                        self.miner_info
+                            .write()
+                            .await
+                            .insert(address, (chrono::Utc::now().timestamp(), share_difficulty, 0));
+                    }
```

Comment:
after having digested this whole block: the [entry API](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry) should be used here instead